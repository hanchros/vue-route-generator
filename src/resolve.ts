import { NestedMap, setToMap } from './nested-map'

export interface PageMeta {
  name: string
  specifier: string
  path: string
  component: string
  children?: PageMeta[]
}

export function resolveRoutePaths(
  paths: string[],
  importPrefix: string
): PageMeta[] {
  const map: NestedMap<string[]> = {}

  const splitted = paths.map(p => p.split('/'))
  splitted.forEach(path => {
    setToMap(map, pathToMapPath(path), path)
  })

  return pathMapToMeta(map, importPrefix)
}

function pathMapToMeta(
  map: NestedMap<string[]>,
  importPrefix: string,
  parentDepth: number = 0
): PageMeta[] {
  if (map.value) {
    const path = map.value
    const meta: PageMeta = {
      name: pathToName(path),
      specifier: pathToSpecifier(path),
      path: pathToRoute(path, parentDepth),
      component: importPrefix + path.join('/')
    }

    if (map.children) {
      meta.children = pathMapChildrenToMeta(
        map.children,
        importPrefix,
        path.length
      )
    }

    return [meta]
  }

  return map.children
    ? pathMapChildrenToMeta(map.children, importPrefix, parentDepth)
    : []
}

function pathMapChildrenToMeta(
  children: Map<string, NestedMap<string[]>>,
  importPrefix: string,
  parentDepth: number
): PageMeta[] {
  return Array.from(children.values()).reduce<PageMeta[]>((acc, value) => {
    return acc.concat(pathMapToMeta(value, importPrefix, parentDepth))
  }, [])
}

function isDynamicRoute(segment: string): boolean {
  return segment[0] === '_'
}

function isOmittable(segment: string): boolean {
  return segment === 'index.vue'
}

function pathToMapPath(segments: string[]): string[] {
  const last = segments[segments.length - 1]
  return segments.slice(0, -1).concat(basename(last))
}

function pathToName(segments: string[]): string {
  const last = segments[segments.length - 1]
  segments = segments.slice(0, -1).concat(basename(last))

  return segments
    .map(s => {
      return s[0] === '_' ? s.slice(1) : s
    })
    .join('-')
}

function pathToSpecifier(segments: string[]): string {
  const name = pathToName(segments)
  const replaced = name
    .replace(/(^|[^a-zA-Z])([a-zA-Z])/g, (_, a, b) => {
      return a + b.toUpperCase()
    })
    .replace(/[^a-zA-Z0-9]/g, '')

  return /^\d/.test(replaced) ? '_' + replaced : replaced
}

function pathToRoute(segments: string[], parentDepth: number): string {
  const last = segments[segments.length - 1]

  if (isOmittable(last)) {
    segments = segments.slice(0, -1)
  } else {
    segments = segments.slice(0, -1).concat(basename(last))
  }

  const routeSegments = segments.slice(parentDepth).map(s => {
    return isDynamicRoute(s) ? ':' + s.slice(1) : s
  })

  const prefix = parentDepth > 0 ? '' : '/'
  return prefix + routeSegments.join('/')
}

function basename(filename: string): string {
  return filename.replace(/\..+?$/, '')
}
