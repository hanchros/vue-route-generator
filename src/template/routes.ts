import { PageMeta } from '../resolve'

function createChildrenRoute(children: PageMeta[]): string {
  return `,
    children: [${children.map(createRoute).join(',')}]`
}

function createRoute(meta: PageMeta): string {
  const children = !meta.children ? '' : createChildrenRoute(meta.children)

  return `
  {
    name: '${meta.name}',
    path: '${meta.path}',
    component: ${meta.specifier}${children}
  }`
}

function createImport(meta: PageMeta, dynamic: boolean): string {
  const code = dynamic
    ? `const ${meta.specifier} = () => import(/* webpackChunkName: "${
        meta.name
      }" */ '${meta.component}')`
    : `import ${meta.specifier} from '${meta.component}'`

  return meta.children
    ? [code]
        .concat(meta.children.map(child => createImport(child, dynamic)))
        .join('\n')
    : code
}

export function createRoutes(meta: PageMeta[], dynamic: boolean): string {
  const imports = meta.map(m => createImport(m, dynamic)).join('\n')
  const code = meta.map(createRoute).join(',')
  return `${imports}\n\nexport default [${code}]`
}
