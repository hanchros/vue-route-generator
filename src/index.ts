import * as fg from 'fast-glob'
import { createRoutes } from './template/routes'
import { resolveRoutePaths } from './resolve'

export interface GenerateConfig {
  pages: string
  importPrefix?: string
  dynamicImport?: boolean
}

export function generateRoutes({
  pages,
  importPrefix = '@/pages/',
  dynamicImport = true
}: GenerateConfig): string {
  const pagePaths = fg.sync<string>('**/*.vue', {
    cwd: pages,
    onlyFiles: true
  })

  const metaList = resolveRoutePaths(pagePaths, importPrefix)

  return createRoutes(metaList, dynamicImport)
}
