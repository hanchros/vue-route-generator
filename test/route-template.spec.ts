import { createRoutes } from '../src/template/routes'
import { PageMeta } from '../src/resolve'

describe('Route template', () => {
  it('should generate routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        component: '@/pages/foo.vue'
      },
      {
        name: 'bar',
        specifier: 'Bar',
        path: '/bar',
        component: '@/pages/bar.vue'
      }
    ]

    expect(createRoutes(meta, true)).toMatchSnapshot()
  })

  it('should generate nested routes', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        component: '@/pages/foo.vue',
        children: [
          {
            name: 'bar',
            specifier: 'FooBar',
            path: 'bar',
            component: '@/pages/bar.vue'
          },
          {
            name: 'baz',
            specifier: 'FooBaz',
            path: 'baz',
            component: '@/pages/baz.vue'
          }
        ]
      }
    ]

    expect(createRoutes(meta, true)).toMatchSnapshot()
  })

  it('should generate static import code', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        component: '@/pages/foo.vue'
      },
      {
        name: 'bar',
        specifier: 'Bar',
        path: '/bar',
        component: '@/pages/bar.vue'
      }
    ]

    expect(createRoutes(meta, false)).toMatchSnapshot()
  })

  it('should generate route meta', () => {
    const meta: PageMeta[] = [
      {
        name: 'foo',
        specifier: 'Foo',
        path: '/foo',
        component: '@/pages/foo.vue',
        routeMeta: {
          title: 'Hello'
        }
      }
    ]

    expect(createRoutes(meta, false)).toMatchSnapshot()
  })
})
