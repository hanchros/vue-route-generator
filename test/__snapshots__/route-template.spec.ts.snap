// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Route template should generate nested routes 1`] = `
"const Foo = () => import(/* webpackChunkName: \\"foo\\" */ '@/pages/foo.vue')
const FooBar = () => import(/* webpackChunkName: \\"bar\\" */ '@/pages/bar.vue')
const FooBaz = () => import(/* webpackChunkName: \\"baz\\" */ '@/pages/baz.vue')

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo,
    children: [
  {
    name: 'bar',
    path: 'bar',
    component: FooBar
  },
  {
    name: 'baz',
    path: 'baz',
    component: FooBaz
  }]
  }]"
`;

exports[`Route template should generate routes 1`] = `
"const Foo = () => import(/* webpackChunkName: \\"foo\\" */ '@/pages/foo.vue')
const Bar = () => import(/* webpackChunkName: \\"bar\\" */ '@/pages/bar.vue')

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo
  },
  {
    name: 'bar',
    path: '/bar',
    component: Bar
  }]"
`;

exports[`Route template should generate static import code 1`] = `
"import Foo from '@/pages/foo.vue'
import Bar from '@/pages/bar.vue'

export default [
  {
    name: 'foo',
    path: '/foo',
    component: Foo
  },
  {
    name: 'bar',
    path: '/bar',
    component: Bar
  }]"
`;
